homeassistant/components/mqtt/__init__.py

```diff
@@ -880,15 +901,19 @@ def _mqtt_on_connect(self, _mqttc, _userdata, _flags, result_code: int) -> None:
             CONF_BIRTH_MESSAGE in self.conf
             and ATTR_TOPIC in self.conf[CONF_BIRTH_MESSAGE]
         ):
-            birth_message = Message(**self.conf[CONF_BIRTH_MESSAGE])
-            self.hass.add_job(
-                self.async_publish(  # pylint: disable=no-value-for-parameter
+
+            async def publish_birth_message(birth_message):
+                await self._ha_started.wait()  # Wait for Home Assistant to start
+                await self._discovery_cooldown()  # Wait for MQTT discovery to cool down
+                await self.async_publish(  # pylint: disable=no-value-for-parameter
                     topic=birth_message.topic,
                     payload=birth_message.payload,
                     qos=birth_message.qos,
                     retain=birth_message.retain,
                 )
-            )
+
+            birth_message = Message(**self.conf[CONF_BIRTH_MESSAGE])
+            self.hass.loop.create_task(publish_birth_message(birth_message))
```

Comment:
In Python 3.7+ we should use `asyncio.create_task` instead. It's a higher level function than the `loop.create_task`.